### PR TITLE
Add FilterFn trait + default implementation

### PR DESCRIPTION
**Pull request by a10y**
Implement `filter` and corresponding `FilterFn` array trait

See #439

### CODE REVIEW

File: vortex-array/src/canonical.rs [hunk 1]
```diff
@@ -48,11 +48,10 @@ use crate::{Array, ArrayDType, IntoArray, ToArray};
 ///
 /// Binary and String views are a new, better encoding format for nearly all use-cases. For now,
 /// because DataFusion does not include pervasive support for compute over StringView, we opt to use
-/// the [`VarBinArray`] as the canonical encoding (which corresponds to the Arrow
-/// [`BinaryViewArray`]).
+/// the [`VarBinArray`] as the canonical encoding (which corresponds to the Arrow `BinaryViewArray`).
 ///
 /// We expect to change this soon once DataFusion is able to finish up some initial support, which
-/// is tracked in https://github.com/apache/datafusion/issues/10918.
+/// is tracked in <https://github.com/apache/datafusion/issues/10918>.
```
Comment (a10y) [inline vortex-array/src/canonical.rs:54]: these were cleanups to make `cargo doc -p vortex-array` complete cleanly

**Review by robert3005 — APPROVED** [review]

File: vortex-array/src/compute/filter.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    );
+    assert_eq!(
+        predicate.len(),
+        array.len(),
+        "predicate.len() must equal array.len()"
+    );
+
+    array.with_dyn(|a| {
+        if let Some(filter_fn) = a.filter() {
+            Ok(filter_fn.filter(array))
```
Comment (joseph-isaacs) [inline vortex-array/src/compute/filter.rs:37]: Should this be `filter_fn.filter(predicate)`

Comment (a10y) [reply vortex-array/src/compute/filter.rs:37]: Whoops, yes

Comment (joseph-isaacs) [reply vortex-array/src/compute/filter.rs:37]: Ill fix it in a follow up.

File: vortex-array/src/compute/filter.rs [hunk 3]
```diff
@@ -0,0 +1,78 @@
+use arrow_array::cast::AsArray;
+use vortex_dtype::{DType, Nullability};
+use vortex_error::VortexResult;
+
+use crate::arrow::FromArrowArray;
+use crate::{Array, ArrayDType, ArrayData, IntoArray, IntoCanonical};
+
+pub trait FilterFn {
+    /// Filter an array by the provided predicate.
+    fn filter(&self, predicate: &Array) -> Array;
```
Comment (joseph-isaacs) [inline vortex-array/src/compute/filter.rs:10]: What do you think about moving to this taking a `&dyn BoolArrayTrait`?

**Pull request merged**